spark/src/main/scala/com/gu/tableversions/spark/PathMapper.scala

```diff
@@ -0,0 +1,76 @@
+package com.gu.tableversions.spark
+
+import java.net.URI
+
+import org.apache.hadoop.fs.Path
+
+// Map paths between one "outer" FileSystem and an underlying one
+// e.g. by changing the schema or appending version directories
+trait PathMapper {
+
+  def forUnderlying(path: Path): Path
+
+  def fromUnderlying(path: Path): Path
+}
+
+class VersionedPathMapper(underlyingFsScheme: String, partitionMappings: Map[String, String]) extends PathMapper {
+
+  // convert a path with a "versioned" scheme to one suitable for the underlying FileSystem:
+  // - replace the versioned schema with the base FS
+  // - append the version directory to the path
+  override def forUnderlying(path: Path): Path = {
+    assert(path.toUri.getScheme == VersionedFileSystem.scheme,
+           s"Path provided to `forUnderlying` ($path) not in the ${VersionedFileSystem.scheme} scheme")
+
+    appendVersion(setUnderlyingScheme(path))
+  }
+
+  // convert a path from the underlying FileSystem to one in the "versioned" scheme:
+  // - set the schema to "versioned"
+  // - remove the version directory from the path (if it exists)
+  override def fromUnderlying(path: Path): Path = {
+    assert(path.toUri.getScheme == underlyingFsScheme,
+           s"Path provided to `fromUnderlying` ($path) not in the underlying $underlyingFsScheme scheme")
+
+    val pathWithoutVersionDirectory = partitionMappings
+      .find {
+        case (_, versionedPartition) =>
+          path.toString.contains(versionedPartition)
+      }
+      .map {
+        case (partition, versionedPartition) =>
+          new Path(path.toString.replace(versionedPartition, partition))
+      }
+      .getOrElse(setVersionedScheme(path))
```

Comment:
Given we set the scheme at the end of this method, I guess we just need `.getOrElse(path)` here?